code/__DATASTRUCTURES/heap.dm

```diff
@@ -0,0 +1,78 @@
+
+//////////////////////
+//Heap object
+//////////////////////
+
+/Heap
+	var/list/L
+	var/cmp
+
+/Heap/New(compare)
+	L = new()
+	cmp = compare
+
+/Heap/proc/IsEmpty()
+	return !L.len
+
+//Insert and place at its position a new node in the heap
+/Heap/proc/Insert(atom/A)
+
+	L.Add(A)
+	Swim(L.len)
+
+//removes and returns the first element of the heap
+//(i.e the max or the min dependant on the comparison function)
+/Heap/proc/Pop()
+	if(!L.len)
+		return 0
+	. = L[1]
+
+	L[1] = L[L.len]
+	L.Cut(L.len)
+
+	Sink(1)
+
+	return .
```

Comment:
remove the ```return``` here, ```return .``` is implied, and forcing it explicitly by writing ```return``` is actually slower.